src/ast/passes/semantic_analyser.cpp

```diff
@@ -2969,45 +2976,53 @@ void SemanticAnalyser::visit(VarDeclStatement &decl)
         << "Invalid variable declaration type: " << decl.var->type;
   }
 
-  // This should be the first time we're seeing this variable
-  if (auto decl_search = variable_decls_[scope_].find(var_ident);
-      decl_search != variable_decls_[scope_].end()) {
-    if (decl_search->second != decl.loc) {
-      // Note this will change if/when proper block/lexical scoping is added
-      LOG(ERROR, decl.loc, err_)
-          << "Variable " << var_ident << " was already declared.";
-      LOG(ERROR, decl_search->second, err_) << "Initial declaration";
+  if (is_first_pass()) {
+    for (auto scope : scope_stack_) {
+      // This should be the first time we're seeing this variable
+      if (auto decl_search = variable_decls_[scope].find(var_ident);
+          decl_search != variable_decls_[scope].end()) {
+        if (decl_search->second != decl.loc) {
+          LOG(ERROR, decl.loc, err_)
+              << "Variable " << var_ident
+              << " was already declared. Variable shadowing is not allowed.";
+          LOG(ERROR, decl_search->second, err_) << "Initial declaration";
+        }
+      }
     }
   }
 
-  if (auto search = variables_[scope_].find(var_ident);
-      search != variables_[scope_].end()) {
-    if (!variable_decls_[scope_].contains(var_ident)) {
-      LOG(ERROR, decl.loc, err_)
-          << "Variable declarations need to occur before variable usage or "
-             "assignment. Variable: "
-          << var_ident;
-    } else if (is_final_pass()) {
-      // Update the declaration type if it was either not set e.g. `let $a;`
-      // or the type is ambiguous or resizable e.g. `let $a: string;`
-      decl.var->type = search->second.type;
-    }
+  if (is_first_pass() || is_final_pass()) {
```

Comment:
Why do this specifically on the first and the last pass? Can we add a comment to explain?

Reply:
The only error checks below are for last or first pass so no need to iterate over the scope stack in other passes. I'm adding a comment as to why we check for `first_pass` - it's similar to above. For cases like this: `BEGIN { if (1) { let $x; } $x = 2; }`